File: frontend/src/lib/matches.ts
/**
 * Resume + matches API client.
 *
 * Hits the FastAPI endpoints we added/patched:
 *   POST /api/resumes/upload       (multipart) → embed + match in one shot
 *   GET  /api/resumes              → list resumes
 *   GET  /api/matches?limit=N      → persisted top matches
 *   POST /api/matches/refresh      → recompute against current active resume
 */
import { authHeaders } from "@/lib/auth";

export interface ResumeRow {
  id: number;
  name: string;
  target_role: string | null;
  is_active: boolean;
  experience_level: string | null;
  overall_strength_score: number | null;
  created_at: string;
}

export interface UploadResponse {
  id: number;
  name: string;
  target_role: string | null;
  skills: string[] | null;
  experience_level: string | null;
  overall_strength_score: number | null;
  is_active: boolean;
  matches_computed: number;
}

export interface MatchRow {
  job_id: number;
  title: string;
  company: string;
  location: string | null;
  remote_type: string | null;
  salary_min: number | null;
  salary_max: number | null;
  url: string | null;
  apply_url: string | null;
  posted_at: string | null;
  first_seen_at: string | null;
  source_ats: string | null;
  match_score: number;
  sim_score: number | null;
  salary_fit: number | null;
  location_fit: number | null;
  freshness_score: number | null;
}

export interface MatchesResponse {
  user_id: number;
  count: number;
  matches: MatchRow[];
}

// ── Resumes ──────────────────────────────────────────────────────────────────

export async function uploadResume(
  file: File,
  targetRole?: string,
): Promise<UploadResponse> {
  const fd = new FormData();
  fd.append("file", file);
  if (targetRole) fd.append("target_role", targetRole);

  const res = await fetch("/api/resumes/upload", {
    method: "POST",
    headers: { ...authHeaders() },
    body: fd,
  });
  const data = await res.json().catch(() => ({}));
  if (!res.ok) {
    throw new Error(data?.detail || `Upload failed: ${res.status}`);
  }
  return data as UploadResponse;
}

export async function listResumes(): Promise<ResumeRow[]> {
  const res = await fetch("/api/resumes", {
    headers: { ...authHeaders() },
  });
  if (!res.ok) throw new Error(`Failed to list resumes: ${res.status}`);
  return res.json();
}

export async function activateResume(
  resumeId: number,
): Promise<{ message: string; matches_computed: number }> {
  const res = await fetch(`/api/resumes/${resumeId}/activate`, {
    method: "PATCH",
    headers: { ...authHeaders() },
  });
  if (!res.ok) throw new Error(`Failed to activate resume: ${res.status}`);
  return res.json();
}

// ── Matches ──────────────────────────────────────────────────────────────────

export type CountryFilter = "us" | "remote" | "all";

export async function getMatches(
  opts: { limit?: number; country?: CountryFilter; recencyDays?: number | null } = {},
): Promise<MatchesResponse> {
  const limit = opts.limit ?? 50;
  const country = opts.country ?? "us";
  const params = new URLSearchParams({
    limit: String(limit),
    country,
  });
  if (opts.recencyDays && opts.recencyDays > 0) {
    params.set("recency_days", String(opts.recencyDays));
  }
  const res = await fetch(`/api/matches?${params}`, {
    headers: { ...authHeaders() },
  });
  if (!res.ok) throw new Error(`Failed to load matches: ${res.status}`);
  return res.json();
}

export async function refreshMatches(top = 50): Promise<{ computed: number }> {
  const res = await fetch(`/api/matches/refresh?top=${top}`, {
    method: "POST",
    headers: { ...authHeaders() },
  });
  if (!res.ok) throw new Error(`Failed to refresh matches: ${res.status}`);
  return res.json();
}

// ── AI assist ────────────────────────────────────────────────────────────────

export async function generateCoverLetter(jobId: number): Promise<{ cover_letter: string }> {
  const r = await fetch(`/api/ai/cover_letter/${jobId}`, {
    method: "POST",
    headers: { ...authHeaders() },
  });
  if (!r.ok) throw new Error((await r.json())?.detail || `Failed: ${r.status}`);
  return r.json();
}

export async function tailorResumeForJob(jobId: number): Promise<{ tailored_resume: string }> {
  const r = await fetch(`/api/ai/tailor_resume/${jobId}`, {
    method: "POST",
    headers: { ...authHeaders() },
  });
  if (!r.ok) throw new Error((await r.json())?.detail || `Failed: ${r.status}`);
  return r.json();
}

export async function autoApply(jobIds: number[], dryRun = true): Promise<{ queued: number }> {
  const r = await fetch("/api/ai/auto_apply", {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({ job_ids: jobIds, dry_run: dryRun }),
  });
  if (!r.ok) throw new Error((await r.json())?.detail || `Failed: ${r.status}`);
  return r.json();
}

File: backend/app/api/v1/matches.py
"""Personalized job matches — driven by resume embedding cosine similarity."""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.core.dependencies import get_current_user
from app.models.user import User
from app.services.match_service import (
    list_matches_for_user,
    recompute_matches_for_user,
)

router = APIRouter(prefix="/matches", tags=["matches"])


def _serialize(m: dict) -> dict:
    """Convert raw row to JSON-friendly dict."""
    return {
        "job_id":          m["id"],
        "title":           m["title"],
        "company":         m.get("company_name") or "",
        "location":        m.get("location"),
        "remote_type":     m.get("remote_type"),
        "salary_min":      m.get("salary_min"),
        "salary_max":      m.get("salary_max"),
        "url":             m.get("url"),
        "apply_url":       m.get("apply_url") or m.get("url"),
        "posted_at":       m["posted_at"].isoformat()    if m.get("posted_at")    else None,
        "first_seen_at":   m["first_seen_at"].isoformat() if m.get("first_seen_at") else None,
        "source_ats":      m.get("source_ats"),
        "match_score":     float(m["match_score"]),
        "sim_score":       float(m["sim_score"])      if m.get("sim_score")      is not None else None,
        "salary_fit":      float(m["salary_fit"])     if m.get("salary_fit")     is not None else None,
        "location_fit":    float(m["location_fit"])   if m.get("location_fit")   is not None else None,
        "freshness_score": float(m["freshness_score"]) if m.get("freshness_score") is not None else None,
    }


@router.get("")
async def get_matches(
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Return the user's persisted top matches (sorted by composite score).
    The matches are populated automatically on resume upload; the user can
    also force a recompute via POST /matches/refresh.
    """
    rows = await list_matches_for_user(db, current_user.id, limit=limit)
    return {
        "user_id": current_user.id,
        "count": len(rows),
        "matches": [_serialize(r) for r in rows],
    }


@router.post("/refresh")
async def refresh_matches(
    top: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Recompute matches for the current user (uses active resume)."""
    n = await recompute_matches_for_user(db, current_user.id, top=top)
    return {"computed": n}
